### ISSUE TITLE
Printing information about GFLOPs at each layer of the YOLOv5 model failed

### ISSUE THREAD

**Issue by xuchuanzhen** [ultralytics/ultralytics#3464]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Other

### Bug

The screenshot of the error is as follows:
![image](https://github.com/ultralytics/ultralytics/assets/131573646/cfe8ebdc-cd12-45ca-b98d-6fa707a9ff93)

```
Traceback (most recent call last):
  File "E:\DL\ultralytics-xcz\ultralytics\flops.py", line 18, in <module>
    model.predict(input, profile=True)
  File "E:\DL\ultralytics-xcz\ultralytics\nn\tasks.py", line 62, in predict
    return self._predict_once(x, profile, visualize)
  File "E:\DL\ultralytics-xcz\ultralytics\nn\tasks.py", line 81, in _predict_once
    self._profile_one_layer(m, x, dt)
  File "E:\DL\ultralytics-xcz\ultralytics\nn\tasks.py", line 109, in _profile_one_layer
    o = thop.profile(m, inputs=[x.clone() if c else x], verbose=False)[0] / 1E9 * 2 if thop else 0  # FLOPs
AttributeError: 'list' object has no attribute 'clone'



```

### Environment

-YOLO: Ultralytics YOLOv8.0.123  Python-3.9.7 torch-2.0.1 CUDA:0 (NVIDIA GeForce RTX 3050ti, 3976MiB)
-Operating system: windows11

### Minimal Reproducible Example

```
import torch

from ultralytics.nn.tasks import  DetectionModel

model = DetectionModel()  # build model
input = torch.randn(32, 3, 640, 640)
model.predict(input, profile=True)
```

### Additional

```
#The _profile_one_layer function in the task.py file reported an error
    def _profile_one_layer(self, m, x, dt):
        """
        Profile the computation time and FLOPs of a single layer of the model on a given input.
        Appends the results to the provided list.

        Args:
            m (nn.Module): The layer to be profiled.
            x (torch.Tensor): The input data to the layer.
            dt (list): A list to store the computation time of the layer.

        Returns:
            None
        """
        c = m == self.model[-1]  # is final layer, copy input as inplace fix
        o = thop.profile(m, inputs=[x.clone() if c else x], verbose=False)[0] / 1E9 * 2 if thop else 0  # FLOPs
        # o = thop.profile(m, inputs=[x], verbose=False)[0] / 1E9 * 2 if thop else 0  # FLOPs
        t = time_sync()
        for _ in range(10):
            m(x.clone() if c else x)
            # m(x)
        dt.append((time_sync() - t) * 100)
        if m == self.model[0]:
            LOGGER.info(f"{'time (ms)':>10s} {'GFLOPs':>10s} {'params':>10s}  module")
        LOGGER.info(f'{dt[-1]:10.2f} {o:10.2f} {m.np:10.0f}  {m.type}')
        if c:
            LOGGER.info(f"{sum(dt):10.2f} {'-':>10s} {'-':>10s}  Total")
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@xuchuanzhen I'm unable to reproduce your issue, instead when I run your code I get this error:

```python
import torch

from ultralytics.nn.tasks import  DetectionModel

model = DetectionModel()  # build model
input = torch.randn(32, 3, 640, 640)
model.predict(input, profile=True)
```

```
ModuleNotFoundError: No module named 'ultralytics.cfg'
```

**Comment by glenn-jocher** [comment]
@xuchuanzhen thanks for the bug report! I'm able to reproduce the issue using your code. I've added a TODO to this bug to prioritize this for a fix in the next few days.

<img width="992" alt="Screenshot 2023-06-30 at 00 01 25" src="https://github.com/ultralytics/ultralytics/assets/26833433/640b9864-3280-41dc-9143-8f512033477f">

**Comment by songjiahao-wq** [comment]
> @xuchuanzhen thanks for the bug report! I'm able to reproduce the issue using your code. I've added a TODO to this bug to prioritize this for a fix in the next few days.
> 
> <img alt="Screenshot 2023-06-30 at 00 01 25" width="992" src="https://user-images.githubusercontent.com/26833433/249906724-640b9864-3280-41dc-9143-8f512033477f.png">
Hello, I also found this problem when printing the number of parameters, and solved it by changing the clone() to copy()
![image](https://github.com/ultralytics/ultralytics/assets/75210061/dd60fc9a-f317-49eb-92ec-a04b10c33d46)

**Comment by xuchuanzhen** [comment]
> > @xuchuanzhen thanks for the bug report! I'm able to reproduce the issue using your code. I've added a TODO to this bug to prioritize this for a fix in the next few days.
> > <img alt="Screenshot 2023-06-30 at 00 01 25" width="992" src="https://user-images.githubusercontent.com/26833433/249906724-640b9864-3280-41dc-9143-8f512033477f.png">
> 
> Hello, I also found this problem when printing the number of parameters, and solved it by changing the clone() to copy() ![image](https://user-images.githubusercontent.com/75210061/252149680-dd60fc9a-f317-49eb-92ec-a04b10c33d46.png)

我在ultralytics/yolo/engine/model.py中移植了profile函数也解决了这个问题
```
    def profile(self, imgsz):
        if type(imgsz) is int:
            inputs = torch.randn((2, 3, imgsz, imgsz))
        else:
            inputs = torch.randn((2, 3, imgsz[0], imgsz[1]))
        return self.model.predict(inputs, profile=True)
```

**Comment by glenn-jocher** [comment]
Hi @xuchuanzhen, thank you for reporting this issue and sharing your solution! We appreciate your contribution. 

We have also identified the problem and have added a TODO to prioritize this bug for a fix in the next few days. In the meantime, @maidor from the Ultralytics team has suggested another workaround by porting the profile function in `ultralytics/yolo/engine/model.py`. 

Here is the code snippet they provided:

```python
def profile(self, imgsz):
    if type(imgsz) is int:
        inputs = torch.randn((2, 3, imgsz, imgsz))
    else:
        inputs = torch.randn((2, 3, imgsz[0], imgsz[1]))
    return self.model.predict(inputs, profile=True)
```

This will perform profiling on your YOLOv8 model. 

Once again, thank you for bringing this issue to our attention and for your contribution! Please let us know if you have any further questions.

**Comment by glenn-jocher** [comment]
@xuchuanzhen great news 😃! Your original issue may now be resolved ✅ in PR #4048. With this PR the following line profiling code now works correctly for all YOLO models:

```python
import torch

from ultralytics.nn.tasks import  DetectionModel

model = DetectionModel()  # build model
input = torch.randn(32, 3, 640, 640)
model.predict(input, profile=True)
```

To get this update:
- **[Git](https://github.com/ultralytics/ultralytics)** – Run `git pull` from within your `ultralytics/` directory or run `git clone https://github.com/ultralytics/ultralytics` again
- **[Pip](https://pypi.org/project/ultralytics/)** – Update with `pip install -U ultralytics`
- **[Notebooks](https://github.com/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb)** – Check out the updated notebooks  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/ultralytics)** – Run `sudo docker pull ultralytics/ultralytics:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and letting us know. Please confirm if this update fixes the issue for you, and don't hesitate to report any other **issues** you find or **feature requests** you may have. Happy training with YOLOv8 🚀!
